### PR TITLE
Implement snapcraft search

### PR DESCRIPTION
**Pull request by sergiusens**
LP: #1596222

Signed-off-by: Sergio Schvezov <sergio.schvezov@ubuntu.com>

### CODE REVIEW

File: snapcraft/internal/common.py [hunk 1]
```diff
@@ -212,6 +212,16 @@ def _search_and_replace_contents(file_path, search_pattern, replacement):
...
 
 
+def get_terminal_width(max_width=MAX_CHARACTERS_WRAP):
+    if os.isatty(sys.stdout.fileno()):
+        width = shutil.get_terminal_size().columns
+    else:
+        width = max_width
+    if max_width:
+        width = min(max_width, width)
+    return width
```
Comment (elopio) [inline snapcraft/internal/common.py]: This seems wrong. 
You are calling: terminal_width = get_terminal_width(max_width=None). For not a tty, that will return None
Then you do: description_space = terminal_width - part_length - 5
So None - something - 5, will raise:
TypeError: unsupported operand type(s) for -: 'NoneType' and 'int'

File: snapcraft/internal/parts.py [hunk 2]
```diff
@@ -117,6 +124,25 @@ def get_part(self, part_name, full=False):
                 remote_part.pop(key)
         return remote_part
 
+    def matches_for(self, part_match, max_len=0):
+        sm = difflib.SequenceMatcher(isjunk=None, autojunk=False)
+        sm.set_seq2(part_match)
```
Comment (elopio) [inline snapcraft/internal/parts.py]: using sm as a variable is a gopher thing. Something like "matcher" would be more pythonic.

File: snapcraft/internal/parts.py [hunk 3]
```diff
@@ -117,6 +124,25 @@ def get_part(self, part_name, full=False):
...
+        matching_parts = {}
+        for part_name in self._parts.keys():
+            sm.set_seq1(part_name)
+            add_part_name = sm.ratio() >= _MATCH_RATIO
+            sm.set_seq1(self._parts[part_name]['description'])
+            add_description = sm.ratio() >= _MATCH_RATIO
+
+            if add_part_name or add_description or part_match in part_name:
+                matching_parts[part_name] = self._parts[part_name]
+                matching_parts[part_name] = self._parts[part_name]
```
Comment (elopio) [inline snapcraft/internal/parts.py]: why twice?

File: snapcraft/internal/parts.py [hunk 4]
```diff
@@ -117,6 +124,25 @@ def get_part(self, part_name, full=False):
                 remote_part.pop(key)
         return remote_part
 
+    def matches_for(self, part_match, max_len=0):
```
Comment (elopio) [inline snapcraft/internal/parts.py]: _matches_for

File: snapcraft/internal/parts.py [hunk 5]
```diff
@@ -151,5 +177,30 @@ def define(part_name):
...
+
+    terminal_width = get_terminal_width(max_width=None)
+    part_length = max(part_length, header_len)
+    # <space> + <space> + <description> + ... = 5
+    description_space = terminal_width - part_length - 5
+
+    if not matches:
+        # apt search does not return error, we probably shouldn't either.
+        logger.info('No matches found, try to run `snapcraft update` to '
+                    'refresh the remote parts cache.')
```
Comment (elopio) [inline snapcraft/internal/parts.py]: or maybe the part just doesn't exist. Should that be part of the message?
What about: to contribute new parts, read <some doc> ?

File: snapcraft/tests/test_commands_search.py [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+                        new=tests.SilentProgressBar):
+            parts.update()
+
+        patcher = mock.patch('snapcraft.internal.parts.get_terminal_width')
+        self.mock_terminal = patcher.start()
+        self.mock_terminal.return_value = 80
+        self.addCleanup(patcher.stop)
+
+    @mock.patch('sys.stdout', new_callable=io.StringIO)
+    def test_searching_for_a_part_that_exists(self, mock_stdout):
```
Comment (elopio) [inline snapcraft/tests/test_commands_search.py]: you are missing a test for a match in description, not in name.
And for completenes, maybe a match in name, not in description.

File: snapcraft/internal/common.py (hunk 1 above)
Comment (sergiusens) [inline snapcraft/internal/common.py]: Ah, nice catch

File: snapcraft/internal/parts.py (hunk 4 above)
Comment (sergiusens) [inline snapcraft/internal/parts.py]: The class is private, the method public, I don't see the reason for this one.

File: snapcraft/internal/parts.py (hunk 3 above)
Comment (sergiusens) [inline snapcraft/internal/parts.py]: this is a vim problem

File: snapcraft/internal/parts.py (hunk 5 above)
Comment (sergiusens) [inline snapcraft/internal/parts.py]: well it really is a search string, not really a part

File: snapcraft/tests/test_commands_search.py (hunk 6 above)
Comment (sergiusens) [inline snapcraft/tests/test_commands_search.py]: I removed matching description for now

**Comment by sergiusens** [comment]
@elopio ok, one more round!

File: snapcraft/internal/parts.py (hunk 4 above)
Comment (kyrofa) [inline snapcraft/internal/parts.py]: This method is only ever used within this class though, no? Do you envision it being used by others?

Comment (elopio) [inline snapcraft/internal/parts.py]: This is how I see it: an internal class starts with _ and is used only in the module. An internal method starts with _ and is used only in the class. So internal classes can have internal methods. Of course, this is how I use internals in python, not a good practice I have seen everywhere. So we can have a discussion to define our rules for internals.

File: snapcraft/internal/parts.py [hunk 7]
```diff
@@ -117,6 +124,22 @@ def get_part(self, part_name, full=False):
...
+    def matches_for(self, part_match, max_len=0):
+        matcher = difflib.SequenceMatcher(isjunk=None, autojunk=False)
+        matcher.set_seq2(part_match)
+
+        matching_parts = {}
+        for part_name in self._parts.keys():
+            matcher.set_seq1(part_name)
+            add_part_name = matcher.ratio() >= _MATCH_RATIO
+
+            if add_part_name or part_match in part_name:
```
Comment (elopio) [inline snapcraft/internal/parts.py]: Use parentheses here to make the rules of precedence clear:
if add_part_name or (part_match in part_name):

File: snapcraft/internal/parts.py (hunk 5 above)
Comment (elopio) [inline snapcraft/internal/parts.py]: ok.

File: snapcraft/internal/parts.py [hunk 8]
```diff
@@ -151,5 +174,30 @@ def define(part_name):
...
+        logger.info('No matches found, try to run `snapcraft update` to '
+                    'refresh the remote parts cache.')
+        return
+
+    print('{}  {}'.format(
+        _HEADER_PART_NAME.ljust(part_length, ' '), _HEADER_DESCRIPTION))
+    for part_key in matches.keys():
+        description = matches[part_key]['description']
+        if len(description) > description_space:
+            description = '{}...'.format(description[0:description_space])
```
Comment (kyrofa) [inline snapcraft/internal/parts.py]: I haven't thought through the difficulty of this, but would it be better to wrap instead? Looking something like this:

```
$ snapcraft search qt
PART NAME          DESCRIPTION
qt5conf            This sets up qt5.conf for projects using qml
                   and other qt5 components that need an
                   Ubuntu standard internal path setup by
                   default like the Ubuntu Core Apps.
mqtt-paho/python2  mqtt-paho for python.
mqtt-paho          mqtt-paho for python.
[...]
```

as opposed to:

```
$ snapcraft search qt
PART NAME          DESCRIPTION
qt5conf            This sets up qt5.conf for projects using...
mqtt-paho/python2  mqtt-paho for python.
mqtt-paho          mqtt-paho for python.
[...]
```

File: snapcraft/tests/test_commands_update.py [hunk 9]
```diff
@@ -61,6 +61,12 @@ def test_update(self):
                 'description': 'test entry for part1',
                 'maintainer': 'none',
             },
+            'long-described-part': {
+                'plugin': 'go',
+                'source': 'http://source.tar.gz',
+                'description': 'this is a repetetive description ' * 3,
+                'maintainer': 'none',
+            },
```
Comment (elopio) [inline snapcraft/tests/test_commands_update.py]: where is this one used?

File: snapcraft/internal/parts.py (hunk 7 above)
Comment (sergiusens) [inline snapcraft/internal/parts.py]: Sure, that's doable

File: snapcraft/tests/test_commands_update.py (hunk 9 above)
Comment (sergiusens) [inline snapcraft/tests/test_commands_update.py]: when searching for everything

File: snapcraft/internal/parts.py (hunk 8 above)
Comment (sergiusens) [inline snapcraft/internal/parts.py]: this is bad for filtering @kyrofa

Comment (kyrofa) [inline snapcraft/internal/parts.py]: Filtering? You mean e.g. piping to grep? So then the way to see the entire description is just to `snapcraft define` it?

File: snapcraft/tests/test_commands_update.py (hunk 9 above)
Comment (elopio) [inline snapcraft/tests/test_commands_update.py]: ah, I see it now. It comes from the fake server.

**Comment by elopio** [comment]
There's a typo: repetetive instead of repetitive. Looks good to me.

File: snapcraft/internal/parts.py (hunk 4 above)
Comment (sergiusens) [inline snapcraft/internal/parts.py]: @elopio I am in agreement. An instance of _RemoteParts is created under the implementation of `search` and calls this method so it is not internal to the class https://github.com/ubuntu-core/snapcraft/pull/608/files#diff-ac0a8a66808d29ef0e343cb310093978R179

Comment (sergiusens) [inline snapcraft/internal/parts.py]: @kyrofa no its not https://github.com/ubuntu-core/snapcraft/pull/608/files#diff-ac0a8a66808d29ef0e343cb310093978R179

File: snapcraft/internal/parts.py (hunk 8 above)
Comment (sergiusens) [inline snapcraft/internal/parts.py]: @kyrofa your idea is not bad, but a bug report would be better so if not `grepping` we present a different layout.

Comment (kyrofa) [inline snapcraft/internal/parts.py]: Yeah no problem, I don't have strong feelings here. I can make a bug if you'd like one.

**Comment by kyrofa** [comment]
:+1: from me.

File: snapcraft/internal/parts.py (hunk 4 above)
Comment (elopio) [inline snapcraft/internal/parts.py]: hum, that is more complex. It's like a sandwich of internal. I'll say you win. :D

**Pull request merged**